File: 3518-smallest-palindromic-rearrangement-ii/3518-smallest-palindromic-rearrangement-ii.java
import java.util.*;

class Solution {
    private static final int MAX = 1_000_001;

    public String smallestPalindrome(String s, int k) {
        Map<Character, Integer> count = new HashMap<>();

        for (char c : s.toCharArray()) {
            count.put(c, count.getOrDefault(c, 0) + 1);
        }

        if (!isPalindromePossible(count)) {
            return "";
        }

        HalfResult result = getHalfCountAndMidLetter(count);
        int[] halfCount = result.halfCount;
        String midLetter = result.midLetter;

        int totalPerm = calculateTotalPermutations(halfCount);
        if (k > totalPerm) {
            return "";
        }

        List<Character> leftHalf = generateLeftHalf(halfCount, k);

        StringBuilder ans = new StringBuilder();

        for (char c : leftHalf) {
            ans.append(c);
        }

        ans.append(midLetter);

        for (int i = leftHalf.size() - 1; i >= 0; i--) {
            ans.append(leftHalf.get(i));
        }

        return ans.toString();
    }

    private boolean isPalindromePossible(Map<Character, Integer> count) {
        int oddCount = 0;

        for (int freq : count.values()) {
            if ((freq & 1) == 1) {
                oddCount++;
            }
        }

        return oddCount <= 1;
    }

    private HalfResult getHalfCountAndMidLetter(Map<Character, Integer> count) {
        int[] halfCount = new int[26];
        String midLetter = "";

        for (Map.Entry<Character, Integer> entry : count.entrySet()) {
            char c = entry.getKey();
            int freq = entry.getValue();

            halfCount[c - 'a'] = freq / 2;

            if ((freq & 1) == 1) {
                midLetter = String.valueOf(c);
            }
        }

        return new HalfResult(halfCount, midLetter);
    }

    private int calculateTotalPermutations(int[] halfCount) {
        return countArrangements(halfCount);
    }

    private List<Character> generateLeftHalf(int[] halfCount, int k) {
        int halfLen = 0;
        for (int freq : halfCount) {
            halfLen += freq;
        }

        List<Character> left = new ArrayList<>();

        for (int pos = 0; pos < halfLen; pos++) {
            for (int i = 0; i < 26; i++) {
                if (halfCount[i] == 0) {
                    continue;
                }

                halfCount[i]--;

                int arrangements = countArrangements(halfCount);

                if (arrangements >= k) {
                    left.add((char) ('a' + i));
                    break;
                } else {
                    k -= arrangements;
                    halfCount[i]++;
                }
            }
        }

        return left;
    }

    private int countArrangements(int[] count) {
        int total = 0;

        for (int x : count) {
            total += x;
        }

        long res = 1;

        for (int freq : count) {
            res *= nCk(total, freq);

            if (res >= MAX) {
                return MAX;
            }

            total -= freq;
        }

        return (int) res;
    }

    private int nCk(int n, int k) {
        long res = 1;

        int limit = Math.min(k, n - k);

        for (int i = 1; i <= limit; i++) {
            res = res * (n - i + 1) / i;

            if (res >= MAX) {
                return MAX;
            }
        }

        return (int) res;
    }

    private static class HalfResult {
        int[] halfCount;
        String midLetter;

        HalfResult(int[] halfCount, String midLetter) {
            this.halfCount = halfCount;
            this.midLetter = midLetter;
        }
    }
}

File: leetcode/3518-smallest-palindromic-rearrangement-ii/3518-smallest-palindromic-rearrangement-ii.ts
/**
 * @param s
 * @param k
 * @returns
 */
function smallestPalindrome(s: string, k: number): string {
    const MAX: number = 1_000_001;

    const count: Map<string, number> = new Map();

    for (const c of s) {
        count.set(c, (count.get(c) ?? 0) + 1);
    }

    if (!isPalindromePossible(count)) {
        return "";
    }

    const { halfCount, midLetter } = getHalfCountAndMidLetter(count);

    const totalPerm: number = calculateTotalPermutations(halfCount);
    if (k > totalPerm) {
        return "";
    }

    const leftHalf: string[] = generateLeftHalf(halfCount, k);

    return (
        leftHalf.join("") +
        midLetter +
        [...leftHalf].reverse().join("")
    );

    function isPalindromePossible(count: Map<string, number>): boolean {
        let oddCount: number = 0;

        for (const freq of count.values()) {
            if (freq % 2 === 1) {
                oddCount++;
            }
        }

        return oddCount <= 1;
    }

    function getHalfCountAndMidLetter(
        count: Map<string, number>
    ): { halfCount: number[]; midLetter: string } {
        const halfCount: number[] = Array(26).fill(0);
        let midLetter: string = "";

        for (const [c, freq] of count.entries()) {
            halfCount[c.charCodeAt(0) - 97] = Math.floor(freq / 2);

            if (freq % 2 === 1) {
                midLetter = c;
            }
        }

        return { halfCount, midLetter };
    }

    function calculateTotalPermutations(halfCount: number[]): number {
        return countArrangements(halfCount);
    }

    function generateLeftHalf(
        halfCount: number[],
        k: number
    ): string[] {
        const halfLen: number = halfCount.reduce((a, b) => a + b, 0);
        const left: string[] = [];

        for (let pos = 0; pos < halfLen; pos++) {
            for (let i = 0; i < 26; i++) {
                if (halfCount[i] === 0) {
                    continue;
                }

                halfCount[i]--;

                const arrangements: number = countArrangements(halfCount);

                if (arrangements >= k) {
                    left.push(String.fromCharCode(97 + i));
                    break;
                } else {
                    k -= arrangements;
                    halfCount[i]++;
                }
            }
        }

        return left;
    }

    function countArrangements(count: number[]): number {
        let total: number = count.reduce((a, b) => a + b, 0);
        let res: number = 1;

        for (const freq of count) {
            res *= nCk(total, freq);

            if (res >= MAX) {
                return MAX;
            }

            total -= freq;
        }

        return res;
    }

    function nCk(n: number, k: number): number {
        let res: number = 1;

        for (let i = 1; i <= Math.min(k, n - k); i++) {
            res = Math.floor((res * (n - i + 1)) / i);

            if (res >= MAX) {
                return MAX;
            }
        }

        return res;
    }
}